--- main.c	2015-04-21 12:58:49.000000000 -0400
+++ main.c.new	2015-12-30 15:34:47.000000000 -0500
@@ -525,6 +525,7 @@ int main(int argc, const char* argv[])
   int exitCode = 0;
   int passedConfig;
   bool printGoodbye = true;
+  FILE* fp=NULL;
   
   // By default all messages go to standard error
   setLogMethodToFile(stderr);
@@ -533,6 +534,16 @@ int main(int argc, const char* argv[])
   printf ("Start %s Version%s (%s)\n", SRX_SERVER_NAME, SRX_SERVER_VERSION, 
           __TIME__);
   passedConfig = setupConfiguration(argc, argv);
+
+  if(config.msgDestFilename)
+  {
+    fp = fopen(config.msgDestFilename, "wt");
+    if(fp)
+      setLogMethodToFile(fp);
+    else
+      LOG(LEVEL_ERROR, "Could not set log file.");
+  }
+
   LOG(LEVEL_DEBUG, "([0x%08X]) > Start Main SRx server thread.", pthread_self());
     
   if ( passedConfig != 1)
@@ -603,5 +614,6 @@ int main(int argc, const char* argv[])
   }
   
   LOG(LEVEL_DEBUG, "([0x%08X]) < Stop Main SRx server thread.", pthread_self());
+  fclose(fp);
   return exitCode;
 }
--- log.c	2015-04-23 14:47:09.000000000 -0400
+++ log.c.new	2015-12-30 15:31:50.000000000 -0500
@@ -89,6 +89,7 @@ static void _writeToFile (LogLevel level
   fprintf(_stream, "%s ", LOG_LEVEL_TEXT[level]);
   vfprintf(_stream, fmt, args);
   fputc('\n', _stream);
+  fflush(_stream);
 }
 
 /**
--- configuration.c.old	2015-04-23 17:58:37.000000000 -0400
+++ configuration.c	2015-12-29 15:36:12.000000000 -0500
@@ -190,7 +190,7 @@ void initConfiguration(Configuration* se
                                           "Configuration filename");
   
   self->verbose  = false;
   self->syncAfterConnEstablished = false;
   self->msgDest = MSG_DEST_STDERR;
   self->msgDestFilename = NULL;
@@ -561,8 +561,8 @@ bool readConfigFile(Configuration* self,
   static config_t cfg; // BZ647 changed to static.
   config_setting_t* sett = NULL;
   const char* strtmp = NULL;
-  bool useSyslog;
-  int boolVal;
+  bool useSyslog=0;
+  int boolVal =0;
   int intVal;
 
   // Added intVal to not have the libconfig library modify the configuration
@@ -585,10 +585,12 @@ bool readConfigFile(Configuration* self,
   // Global & server settings
   (void)config_lookup_bool(&cfg, "verbose", (int*)&boolVal);
         self->verbose = (bool)boolVal;
+        boolVal = 0;
   (void)config_lookup_int(&cfg, "port", (int*)&intVal);
         self->server_port = intVal;
   (void)config_lookup_bool(&cfg, "sync", (int*)&boolVal);
         self->syncAfterConnEstablished = (bool)boolVal;
+        boolVal = 0;
 
   (void)config_lookup_int(&cfg, "keep-window", (int*)&intVal);
         self->defaultKeepWindow = intVal;
@@ -596,6 +598,7 @@ bool readConfigFile(Configuration* self,
   // Global - message destination
   (void)config_lookup_bool(&cfg, "syslog", (int*)&boolVal);
         useSyslog = (bool)boolVal;
+        boolVal = 0;
   (void)config_lookup_int(&cfg, "loglevel", (int*)&intVal);
         self->loglevel = intVal;
   if (config_lookup_string(&cfg, "log", &strtmp) == CONFIG_TRUE)
@@ -689,8 +692,10 @@ bool readConfigFile(Configuration* self,
   {
     (void)config_setting_lookup_bool(sett, "no-sendqueue", (int*)&boolVal);
           self->mode_no_sendqueue = (bool)boolVal;
+        boolVal = 0;
     (void)config_setting_lookup_bool(sett, "no-receivequeue", (int*)&boolVal);
-          self->mode_no_receivequeue - (bool)boolVal;
+          self->mode_no_receivequeue = (bool)boolVal;
+        boolVal = 0;
   }
 
   // mapping configuration
