diff --git a/m4/srx_m4_check_openssl.m4 b/m4/srx_m4_check_openssl.m4
index 1cb46ef..f75f03e 100644
--- a/m4/srx_m4_check_openssl.m4
+++ b/m4/srx_m4_check_openssl.m4
@@ -77,17 +77,18 @@ AC_DEFUN([SRX_M4_CHECK_OPENSSL], [
         # libs will be set below in general section
       else
         AC_MSG_RESULT([not found])
+        AC_MSG_CHECKING([for library(static) $srx_m4_libName])
         if test -e $srx_m4_libPath/lib$srx_m4_libName.a; then
+          srx_m4_cflags="-I${srx_m4_openssl_dir}/include"
+          srx_m4_ldflags="-L${srx_m4_libPath}"
+          AC_MSG_RESULT([-l$srx_m4_libName])
+        else
           AC_MSG_ERROR([
     ---------------------------------------------------
     Custom OpenSSL must be configured as shared library
     (config shared ....) to generate lib$srx_m4_libName.so!
     ---------------------------------------------------])
         fi
-        AC_MSG_ERROR([
-    --------------------------------------------------
-    Library $srx_m4_libName required!
-    --------------------------------------------------])
       fi
     fi

@@ -111,6 +112,10 @@ AC_DEFUN([SRX_M4_CHECK_OPENSSL], [

   echo Test for header files

+    
+  TMP_CFLAGS="$srx_m4_cflags"
+  ORIG_CFLAGS=${CFLAGS}
+  CFLAGS=${TMP_CFLAGS}
   AC_CHECK_HEADERS([$srx_m4_openssl_dir/include/openssl/crypto.h \
                     $srx_m4_openssl_dir/include/openssl/bio.h \
                     $srx_m4_openssl_dir/include/openssl/sha.h \
@@ -123,6 +128,7 @@ AC_DEFUN([SRX_M4_CHECK_OPENSSL], [
     not be located.
     --------------------------------------------------]) ])

+  CFLAGS=${ORIG_CFLAGS}
   # Check for libs var
   if test "x$2" = "x"; then
     LIBS="$srx_m4_libs ${LIBS}"
