diff --git AS50/srx/src/util/client_socket.c AS50/srx/src/util/client_socket.c
index e712f77..5d544c4 100644
--- AS50/srx/src/util/client_socket.c
+++ AS50/srx/src/util/client_socket.c
@@ -152,7 +152,8 @@ bool reconnectToServer(ClientSocket* self, int delay, int max_attempts)
     {
       // No more attempts necessary
       max_attempts = 0;
-      continue;
+      //continue;
+      return succ;
     }
 
     // Close the socket - it became invalid
@@ -168,6 +169,7 @@ bool reconnectToServer(ClientSocket* self, int delay, int max_attempts)
     }
   } // while (self->reconnect);
 
+  self->clientFD = -1; // it must be added, othrewise it goes through 'sendResetQuery' that makes SIGPIPE
   return succ;
 }
 
